qiskit/passmanager/base_optimization_tasks.py

```diff
@@ -0,0 +1,231 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Baseclasses for the Qiskit passmanager optimization tasks."""
+from __future__ import annotations
+
+import logging
+import time
+from abc import abstractmethod, ABC
+from collections.abc import Iterable, Callable, Iterator, Sequence
+from typing import Any
+
+from .exceptions import PassManagerError
+from .propertyset import PassState, PropertySet, FencedPropertySet, RunState
+
+logger = logging.getLogger(__name__)
+
+
+class OptimizerTask(ABC):
+    """A definition of optimization task.
+
+    The optimizer task takes a passmanager IR, and outputs new passmanager IR
+    after program optimization. Optimization task can rely on the :class:`.PassState`
+    to communicate intermediate state with other tasks.
+    """
+
+    _state: PassState | None
+
+    @property
+    def state(self) -> PassState:
+        """A local state information associated with this optimization workflow."""
+        if self._state is None:
+            # Allocate state information as needed basis.
+            # This slightly reduces memory footprint for instantiation.
+            self._state = PassState()
+        return self._state
+
+    @property
+    def property_set(self) -> PropertySet:
+        """Property set of this flow controller."""
+        return self.state.property_set
+
+    @property_set.setter
+    def property_set(self, new_property_set: PropertySet):
+        self.state.property_set = new_property_set
```

Comment:
This is good question. This was because I was on conservative side. In principle we can put everything in the `PropertySet`. These values have been instance variables of `RunningPassManager`. Since this object was dropped from the framework, I needed some object to bundle them. There is no technical reason to introduce another `PassStatus` class.

Reply:
I'm fine with separating them. Done in [305fa58](https://github.com/Qiskit/qiskit/pull/10127/commits/305fa589e22ab2f6fd138457f864196ff4a39d5f)